[JIT] Add argumentValue functions and is_mutable checks to SchemaInfo

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81352
* #81007
* #81002
* #81000
* #80984
* __->__ #80972
* #80918

- Created addArgumentValue/s methods in SchemaInfo to pass argument values into the subclass. These are used for more accurate mutation, aliasing and determinism checks which include special cases.
- Added input_alias_map_ to keep track of which inputs alias each other. This is updated with the method generateAliasMap.
- Implemented is_mutable methods in SchemaInfo which also give information based on argument values. For instance, if two inputs alias and one is mutable by the schema, then the other will also be mutable.
- Tested Schema Info is_mutable implementation where inputs alias as mentioned above.

Commit:
[JIT] Add argumentValue functions and is_mutable checks to SchemaInfo

[ghstack-poisoned]

Commit:
Update on "[JIT] Add argumentValue functions and is_mutable checks to SchemaInfo"




[ghstack-poisoned]

Commit:
Update on "[JIT] Add argumentValue functions and is_mutable checks to SchemaInfo"




[ghstack-poisoned]

Commit:
Update on "[JIT] Add argumentValue functions and is_mutable checks to SchemaInfo"




[ghstack-poisoned]

Gamrix commented on torch/csrc/utils/schema_info.h:
```diff
@@ -15,10 +16,38 @@ namespace utils {
+      : FunctionSchema(torch::jit::getOperatorForLiteral(signature)->schema()),
+        updated_(false) {}
+
+  bool is_mutable();
+
+  bool is_mutable(size_t index);
+
+  bool is_mutable(c10::string_view name);
+
```
Note that these will need to be marked as `override` and the base class functions marked as `virtual` for this subclass to be used as intended.

Commit:
Update on "[JIT] Add argumentValue functions and is_mutable checks to SchemaInfo"


- Created addArgumentValue/s methods in SchemaInfo to pass argument values into the subclass. These are used for more accurate mutation, aliasing and determinism checks which include special cases.
- Added input_alias_map_ to keep track of which inputs alias each other. This is updated with the method generateAliasMap.
- Implemented is_mutable methods in SchemaInfo which also give information based on argument values. For instance, if two inputs alias and one is mutable by the schema, then the other will also be mutable.
- Tested Schema Info is_mutable implementation where inputs alias as mentioned above.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add argumentValue functions and is_mutable checks to SchemaInfo"


- Created addArgumentValue/s methods in SchemaInfo to pass argument values into the subclass. These are used for more accurate mutation, aliasing and determinism checks which include special cases.
- Added input_alias_map_ to keep track of which inputs alias each other. This is updated with the method generateAliasMap.
- Implemented is_mutable methods in SchemaInfo which also give information based on argument values. For instance, if two inputs alias and one is mutable by the schema, then the other will also be mutable.
- Tested Schema Info is_mutable implementation where inputs alias as mentioned above.

[ghstack-poisoned]

davidberard98 reviewed (commented):
Looks good for the most part! Requested some minor changes, please address those before merging

goldenxuett commented on test/cpp/jit/test_schema_info.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <gtest/gtest.h>
+#include <torch/csrc/autograd/generated/variable_factories.h>
```
This is used for at::randn()

Commit:
Update on "[JIT] Add argumentValue functions and is_mutable checks to SchemaInfo"


- Created addArgumentValue/s methods in SchemaInfo to pass argument values into the subclass. These are used for more accurate mutation, aliasing and determinism checks which include special cases.
- Added input_alias_map_ to keep track of which inputs alias each other. This is updated with the method generateAliasMap.
- Implemented is_mutable methods in SchemaInfo which also give information based on argument values. For instance, if two inputs alias and one is mutable by the schema, then the other will also be mutable.
- Tested Schema Info is_mutable implementation where inputs alias as mentioned above.

[ghstack-poisoned]

Gamrix commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -1,5 +1,83 @@
+void SchemaInfo::addArgumentValues(
+    const std::vector<c10::optional<at::IValue>>& value_list) {
+  for (size_t i = 0; i < value_list.size(); i++) {
+    if (i < schema_.arguments().size() && value_list[i] != c10::nullopt) {
```
Should `i < schema_.arguments().size()` be a `TORCH_INTERNAL_ASSERT` instead?

Gamrix commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -1,5 +1,83 @@
+  updated_ = true;
+  input_alias_map_ = std::vector<std::unordered_set<size_t>>(
+      schema_.arguments().size(), std::unordered_set<size_t>());
+  for (size_t i = 0; i < schema_.arguments().size(); i++) {
+    for (size_t j = i; j < schema_.arguments().size(); j++) {
```
You should skip arguments that are not tensors.

goldenxuett replied:
Is the reason for that for performance?
Also wouldn't we be worried about any possible container that contains a tensor (like List[List[Tensor]], Dict(Tensor) etc) in case future ops use these containers even if the current ops in native_functions.yaml do not? 
If that's the case, wouldn't the performance decrease from checking whether an IValue is an any level container around Tensor result in around the same level of performance as just considering every type? (I wouldn't be surprised if I'm missing some logic here about this)

Gamrix replied:
You're right, even with 10 arguments, the double for loop isn't that slow.

Commit:
Update on "[JIT] Add argumentValue functions and is_mutable checks to SchemaInfo"


- Created addArgumentValue/s methods in SchemaInfo to pass argument values into the subclass. These are used for more accurate mutation, aliasing and determinism checks which include special cases.
- Added input_alias_map_ to keep track of which inputs alias each other. This is updated with the method generateAliasMap.
- Implemented is_mutable methods in SchemaInfo which also give information based on argument values. For instance, if two inputs alias and one is mutable by the schema, then the other will also be mutable.
- Tested Schema Info is_mutable implementation where inputs alias as mentioned above.

[ghstack-poisoned]

goldenxuett commented:
@pytorchbot merge
